src/os/bluestore/BlueStore.cc

```diff
@@ -3194,94 +3222,108 @@ int BlueStore::_do_read(
   o->flush();
   _dump_onode(o);
 
-  auto lext = lextents.upper_bound(offset);
-  uint32_t l = length;
-  uint64_t off = offset;
-  if (lext == lextents.begin() && offset + length <= lext->first) {
-    bl.append_zero(length);
-    return length;
-  } else if (lext == lextents.begin()) {
-    off = lext->first;
-    l -= lext->first - offset;
-  } else {
-    --lext;
-  }
-
-  ready_regions_t ready_regions_in_cache, ready_regions;
-  interval_set<uint64_t> ready_intervals_in_cache;
-  o->bc.read(off, l, ready_regions_in_cache, ready_intervals_in_cache);
-  dout(20) << __func__ << " regions in cache 0x" << std::hex
-	   << ready_regions_in_cache.size() << " " << ready_intervals_in_cache
-	   << std::dec << dendl;
+  ready_regions_t ready_regions;
 
-  //build blob list to read
+  // build blob-wise list to of stuff read (that isn't cached)
   blobs2read_t blobs2read;
-  while (l > 0 && lext != lextents.end()) {
-    //bluestore_blob_t* bptr = get_blob(lext->second.blob);
-    const bluestore_blob_t *bptr = c->get_blob_ptr(o, lext->second.blob);
+  unsigned left = length;
+  uint64_t pos = offset;
+  auto lp = o->onode.seek_lextent(offset);
+  while (left > 0 && lp != o->onode.extent_map.end()) {
+    if (pos < lp->first) {
+      unsigned hole = lp->first - pos;
+      if (hole >= left) {
+	break;
+      }
+      dout(30) << __func__ << "  hole 0x" << std::hex << pos << "~" << hole
+	       << std::dec << dendl;
+      pos += hole;
+      left -= hole;
+      continue;
+    }
+    Blob *bptr = c->get_blob(o, lp->second.blob);
     if (bptr == nullptr) {
-      dout(20) << __func__ << " missed blob " << lext->second.blob << dendl;
+      dout(20) << __func__ << "  missed blob " << lp->second.blob << dendl;
       _dump_onode(o);
     }
     assert(bptr != nullptr);
-    uint32_t l2read;
-    if (off >= lext->first && off < lext->first + lext->second.length) {
-      uint32_t r_off = off - lext->first;
-      l2read = MIN(l, lext->second.length - r_off);
-      if (!ready_intervals_in_cache.contains( off, l2read)) {
-	regions2read_t& regions = blobs2read[bptr];
-	regions.push_back(region_t(off, r_off + lext->second.offset, l2read));
+    unsigned l_off = pos - lp->first;
+    unsigned b_off = l_off + lp->second.offset;
+    unsigned b_len = std::min(std::min(left,
+				       lp->second.length - l_off),
+			      bptr->blob.length - b_off);
```

Comment:
Not sure we need to use bptr->blob.length - b_off as an additional limitation. In a consistent state lp->second.length - l_off <= ptr->blob.length - b_off. Probably appropriate assert is enough.